File: app/src/main/java/danb/speedrunbrowser/utils/AutoCompleteAdapter.java
package danb.speedrunbrowser.utils;

import android.content.Context;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import java.net.MalformedURLException;
import java.net.URL;
import java.util.ArrayList;
import java.util.LinkedList;
import java.util.List;
import java.util.ListIterator;
import java.util.concurrent.TimeUnit;

import androidx.annotation.NonNull;
import danb.speedrunbrowser.R;
import danb.speedrunbrowser.api.SpeedrunMiddlewareAPI;
import danb.speedrunbrowser.api.objects.SearchResultItem;
import danb.speedrunbrowser.api.objects.User;
import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.CompositeDisposable;
import io.reactivex.functions.Consumer;
import io.reactivex.functions.Function;
import io.reactivex.functions.Predicate;
import io.reactivex.subjects.Subject;

public class AutoCompleteAdapter extends BaseAdapter implements Consumer<SpeedrunMiddlewareAPI.APISearchResponse> {
    private static final String TAG = AutoCompleteAdapter.class.getSimpleName();

    private static final int DEBOUNCE_SEARCH_DELAY = 500;

    private Context ctx;

    private CompositeDisposable disposables;

    private String query;

    private SpeedrunMiddlewareAPI.APISearchData rawSearchData;

    private List<SearchResultItem> searchResults;

    public AutoCompleteAdapter(@NonNull Context context, CompositeDisposable disposables) {
        ctx = context;
        this.disposables = disposables;
        searchResults = new ArrayList<>();
    }

    public void setSearchQuery(String q) {
        this.query = q.toLowerCase();
    }

    public void recalculateSearchResults() {
        searchResults = new LinkedList<>();

        searchResults.addAll(rawSearchData.getGames());

        if(!searchResults.isEmpty()) {
            ListIterator<SearchResultItem> sr = searchResults.listIterator();
            SearchResultItem cur = sr.next();
            for(User player : rawSearchData.getPlayers()) {
                // select the longest matching substring
                LCSMatcher lcsp = new LCSMatcher(query, player.getResolvedName().toLowerCase(), 3);

                LCSMatcher lcsg;
                do {
                    lcsg = new LCSMatcher(query, cur.getResolvedName().toLowerCase(), 3);
                    System.out.println("LCSG (" + cur.getResolvedName() + ", " + player.getResolvedName() + "): " + lcsg.getMaxMatchLength() + ", " + lcsp.getMaxMatchLength());
                } while(lcsg.getMaxMatchLength() >= lcsp.getMaxMatchLength() && sr.hasNext() && (cur = sr.next()) != null);

                sr.previous();
                sr.add(player);
                sr.next();
            }
        }

        searchResults.addAll(rawSearchData.getPlayers());

        searchResults = new ArrayList<>(searchResults);

        notifyDataSetInvalidated();
    }

    @Override
    public int getCount() {
        return searchResults.size();
    }

    @Override
    public Object getItem(int position) {
        return searchResults.get(position);
    }

    @Override
    public long getItemId(int position) {
        return 0;
    }

    @Override
    public View getView(int position, View convertView, ViewGroup parent) {
        if(convertView == null)
            convertView = ((LayoutInflater)ctx.getSystemService(Context.LAYOUT_INFLATER_SERVICE)).inflate(R.layout.content_named_autocomplete_item, parent, false);

        SearchResultItem item = searchResults.get(position);

        ImageView viewIcon = convertView.findViewById(R.id.imgItemIcon);
        LinearLayout viewName = convertView.findViewById(R.id.txtItemName);
        TextView viewType = convertView.findViewById(R.id.txtItemType);

        URL iconUrl;
        if((iconUrl = item.getIconUrl()) != null) {
            disposables.add(new ImageLoader(ctx).loadImage(iconUrl)
                    .subscribe(new ImageViewPlacerConsumer(viewIcon)));
        }

        viewName.removeAllViews();

        TextView tv = new TextView(ctx);
        item.applyTextView(tv);
        viewName.addView(tv);

        viewType.setText(item.getType());

        return convertView;
    }

    public void setPublishSubject(Subject<String> subj) {

        Observable<String> obs = subj
                .distinctUntilChanged()
                .filter(new Predicate<String>() {
                    @Override
                    public boolean test(String s) throws Exception {
                        return s != null && (s.isEmpty() || s.length() >= SpeedrunMiddlewareAPI.INSTANCE.getMIN_AUTOCOMPLETE_LENGTH());
                    }
                });

        disposables.add(obs.subscribe(new Consumer<String>() {
            @Override
            public void accept(String s) throws Exception {
                query = s;
            }
        }));

        disposables.add(obs
            .throttleLatest(DEBOUNCE_SEARCH_DELAY, TimeUnit.MILLISECONDS)
            .switchMap(new Function<String, ObservableSource<SpeedrunMiddlewareAPI.APISearchResponse>>() {
                @Override
                public ObservableSource<SpeedrunMiddlewareAPI.APISearchResponse> apply(String s) throws Exception {
                    if(s.length() < SpeedrunMiddlewareAPI.INSTANCE.getMIN_AUTOCOMPLETE_LENGTH())
                        return Observable.just(new SpeedrunMiddlewareAPI.APISearchResponse());
                    else
                        return SpeedrunMiddlewareAPI.INSTANCE.make().autocomplete(s);
                }
            })
            .observeOn(AndroidSchedulers.mainThread())
            .subscribe(this));
    }

    @Override
    public void accept(SpeedrunMiddlewareAPI.APISearchResponse apiSearchResponse) {
        // TODO: Handle error

        rawSearchData = apiSearchResponse.getSearch();
        recalculateSearchResults();
    }
}

File: app/src/main/java/danb/speedrunbrowser/utils/AutoCompleteAdapter.kt
package danb.speedrunbrowser.utils

import android.content.Context
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.BaseAdapter
import android.widget.ImageView
import android.widget.LinearLayout
import android.widget.TextView

import java.util.ArrayList
import java.util.LinkedList
import java.util.concurrent.TimeUnit
import danb.speedrunbrowser.R
import danb.speedrunbrowser.api.SpeedrunMiddlewareAPI
import danb.speedrunbrowser.api.objects.Game
import danb.speedrunbrowser.api.objects.SearchResultItem
import danb.speedrunbrowser.api.objects.User
import io.reactivex.Observable
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.CompositeDisposable
import io.reactivex.functions.Consumer
import io.reactivex.subjects.Subject

class AutoCompleteAdapter(private val ctx: Context, private val disposables: CompositeDisposable) : BaseAdapter(), Consumer<SpeedrunMiddlewareAPI.APISearchResponse> {

    var query: String = ""

    private var rawSearchData: SpeedrunMiddlewareAPI.APISearchData? = null

    private var searchResults: MutableList<SearchResultItem>? = null

    init {
        searchResults = ArrayList()
    }

    private fun mergeSearchResults(results: List<SearchResultItem>) {
        if (searchResults!!.isNotEmpty()) {
            val sr = searchResults!!.listIterator()
            var cur = sr.next()
            for (resultItem in results) {
                // select the longest matching substring
                val lcsp = LCSMatcher(query.toLowerCase(), resultItem.resolvedName.toLowerCase(), 3)

                var lcsg: LCSMatcher
                do {
                    lcsg = LCSMatcher(query.toLowerCase(), cur.resolvedName.toLowerCase(), 3)

                    if(sr.hasNext())
                        cur = sr.next()
                } while (lcsg.maxMatchLength >= lcsp.maxMatchLength && lcsp.maxMatchLength < resultItem.resolvedName.length - 1 && sr.hasNext())

                sr.previous()
                if(sr.hasPrevious())
                    sr.previous()
                sr.add(resultItem)
                sr.next()
            }
        }

        searchResults!!.addAll(results)
    }

    private fun recalculateSearchResults() {
        searchResults = LinkedList()

        searchResults!!.addAll(rawSearchData!!.games)
        mergeSearchResults(rawSearchData!!.players)
        mergeSearchResults(rawSearchData!!.game_groups)

        searchResults = ArrayList(searchResults!!)

        notifyDataSetChanged()
    }

    override fun getCount(): Int {
        return searchResults!!.size
    }

    override fun getItem(position: Int): Any {
        return searchResults!![position]
    }

    override fun getItemId(position: Int): Long {
        return 0
    }

    override fun getView(position: Int, convertView: View?, parent: ViewGroup): View {
        var convertView = convertView
        if (convertView == null)
            convertView = (ctx.getSystemService(Context.LAYOUT_INFLATER_SERVICE) as LayoutInflater).inflate(R.layout.content_named_autocomplete_item, parent, false)

        val item = searchResults!![position]

        val viewIcon = convertView!!.findViewById<ImageView>(R.id.imgItemIcon)
        val viewName = convertView.findViewById<LinearLayout>(R.id.txtItemName)
        val viewType = convertView.findViewById<TextView>(R.id.txtItemType)

        val iconUrl = item.iconUrl
        if (iconUrl != null) {
            disposables.add(ImageLoader(ctx).loadImage(iconUrl)
                    .subscribe(ImageViewPlacerConsumer(viewIcon)))
        }
        else
            viewIcon.setImageDrawable(null)

        viewName.removeAllViews()

        val tv = TextView(ctx)
        item.applyTextView(tv)
        viewName.addView(tv)

        viewType.text = item.type

        return convertView
    }

    fun setPublishSubject(subj: Subject<String>) {

        val obs = subj
                .distinctUntilChanged()
                .filter { it.isEmpty() || it.length >= SpeedrunMiddlewareAPI.MIN_AUTOCOMPLETE_LENGTH }

        disposables.add(obs.subscribe { s -> query = s })

        disposables.add(obs
                .throttleLast(DEBOUNCE_SEARCH_DELAY.toLong(), TimeUnit.MILLISECONDS)
                .switchMap<SpeedrunMiddlewareAPI.APISearchResponse> { s ->
                    if (s.length < SpeedrunMiddlewareAPI.MIN_AUTOCOMPLETE_LENGTH)
                        Observable.just(SpeedrunMiddlewareAPI.APISearchResponse())
                    else
                        SpeedrunMiddlewareAPI.make().autocomplete(s)
                }
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(this))
    }

    override fun accept(apiSearchResponse: SpeedrunMiddlewareAPI.APISearchResponse) {
        // TODO: Handle error

        rawSearchData = apiSearchResponse.search
        recalculateSearchResults()
    }

    companion object {
        private val TAG = AutoCompleteAdapter::class.java.simpleName

        private val DEBOUNCE_SEARCH_DELAY = 200
    }
}
